pcsx2/Frontend/SDLInputSource.cpp

```diff
@@ -0,0 +1,513 @@
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "PrecompiledHeader.h"
+#include "Frontend/SDLInputSource.h"
+#include "Frontend/InputManager.h"
+#include "Host.h"
+#include "HostSettings.h"
+#include "common/Assertions.h"
+#include "common/StringUtil.h"
+#include "common/Console.h"
+#include <cmath>
+
+static const char* s_sdl_axis_names[] = {
+	"LeftX", // SDL_CONTROLLER_AXIS_LEFTX
+	"LeftY", // SDL_CONTROLLER_AXIS_LEFTY
+	"RightX", // SDL_CONTROLLER_AXIS_RIGHTX
+	"RightY", // SDL_CONTROLLER_AXIS_RIGHTY
+	"LeftTrigger", // SDL_CONTROLLER_AXIS_TRIGGERLEFT
+	"RightTrigger", // SDL_CONTROLLER_AXIS_TRIGGERRIGHT
+};
+
+static const char* s_sdl_button_names[] = {
+	"A", // SDL_CONTROLLER_BUTTON_A
+	"B", // SDL_CONTROLLER_BUTTON_B
+	"X", // SDL_CONTROLLER_BUTTON_X
+	"Y", // SDL_CONTROLLER_BUTTON_Y
+	"Back", // SDL_CONTROLLER_BUTTON_BACK
+	"Guide", // SDL_CONTROLLER_BUTTON_GUIDE
+	"Start", // SDL_CONTROLLER_BUTTON_START
+	"LeftStick", // SDL_CONTROLLER_BUTTON_LEFTSTICK
+	"RightStick", // SDL_CONTROLLER_BUTTON_RIGHTSTICK
+	"LeftShoulder", // SDL_CONTROLLER_BUTTON_LEFTSHOULDER
+	"RightShoulder", // SDL_CONTROLLER_BUTTON_RIGHTSHOULDER
+	"DPadUp", // SDL_CONTROLLER_BUTTON_DPAD_UP
+	"DPadDown", // SDL_CONTROLLER_BUTTON_DPAD_DOWN
+	"DPadLeft", // SDL_CONTROLLER_BUTTON_DPAD_LEFT
+	"DPadRight", // SDL_CONTROLLER_BUTTON_DPAD_RIGHT
+	"Misc1", // SDL_CONTROLLER_BUTTON_MISC1
+	"Paddle1", // SDL_CONTROLLER_BUTTON_PADDLE1
+	"Paddle2", // SDL_CONTROLLER_BUTTON_PADDLE2
+	"Paddle3", // SDL_CONTROLLER_BUTTON_PADDLE3
+	"Paddle4", // SDL_CONTROLLER_BUTTON_PADDLE4
+	"Touchpad", // SDL_CONTROLLER_BUTTON_TOUCHPAD
+};
+
+SDLInputSource::SDLInputSource() = default;
+
+SDLInputSource::~SDLInputSource() { pxAssert(m_controllers.empty()); }
+
+bool SDLInputSource::Initialize(SettingsInterface& si)
+{
+	std::optional<std::vector<u8>> controller_db_data = Host::ReadResourceFile("game_controller_db.txt");
+	if (controller_db_data.has_value())
+	{
+		SDL_RWops* ops = SDL_RWFromConstMem(controller_db_data->data(), static_cast<int>(controller_db_data->size()));
+		if (SDL_GameControllerAddMappingsFromRW(ops, true) < 0)
+			Console.Error("SDL_GameControllerAddMappingsFromRW() failed: %s", SDL_GetError());
+	}
+	else
+	{
+		Console.Error("Controller database resource is missing.");
+	}
+
+	const bool ds4_rumble_enabled = si.GetBoolValue("InputSources", "SDLControllerEnhancedMode", false);
+	if (ds4_rumble_enabled)
+	{
+		Console.WriteLn("Enabling PS4/PS5 enhanced mode.");
+#if SDL_VERSION_ATLEAST(2, 0, 9)
+		SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_PS4, "true");
+		SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_PS4_RUMBLE, "true");
+#endif
+#if SDL_VERSION_ATLEAST(2, 0, 16)
+		SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_PS5, "true");
+		SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_PS5_RUMBLE, "true");
+#endif
```

Comment:
Can we include a method to set hints with a configuration file, [like we do on current PCSX2](https://github.com/PCSX2/pcsx2/blob/e5c4fe003b9cfd1821fe81ae48f7d5069d75e865/pcsx2/PAD/Linux/SDL/joystick.cpp#L46-L47)?

Also, can you set these using strings so that they'll automatically be enabled if the user swaps out the sdl dylib for a newer one without recompiling?

Also, please set the `SDL_GAMECONTROLLER_USE_BUTTON_LABELS=0` and `SDL_JOYSTICK_HIDAPI_SWITCH_HOME_LED=0` hints as well.

Side note: We currently require SDL 2.0.9.  You're welcome to lower that, but don't feel like you have to

Reply:
I'm planning on static linking SDL2 (with only the controller support enabled), so this won't be a problem.

Options can be added later when I finish the controller settings UI.